src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs

```diff
@@ -0,0 +1,131 @@
+        // internal for testing.
+        internal List<object> Entries { get; set; } = new List<object>();
+
+        /// <summary>
+        /// Appends the string to the collection.
+        /// </summary>
+        /// <param name="value">The <c>string</c> to be appended.</param>
+        public void Append([NotNull] string value)
+        {
+            Entries.Add(value);
+        }
+
+        /// <summary>
+        /// Appends a character array to the collection.
+        /// </summary>
+        /// <param name="value">The character array to be appended.</param>
+        /// <param name="index">The index from which the character array must be read.</param>
+        /// <param name="count">The count till which the character array must be read.</param>
+        /// <remarks>
+        /// Splits the character array into strings of 1KB length and appends them.
+        /// </remarks>
+        public void Append([NotNull] char[] value, int index, int count)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            if (count < 0 || value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = count;
+                if (MaxCharToStringLength < currentCount)
+                {
+                    currentCount = MaxCharToStringLength;
+                }
+
+                Append(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+        }
+
+        /// <summary>
+        /// Appends a <see cref="IHtmlContent"/> to the collection.
+        /// </summary>
+        /// <param name="htmlContent">The <see cref="IHtmlContent"/> to be appended.</param>
+        public void Append([NotNull] IHtmlContent htmlContent)
+        {
+            Entries.Add(htmlContent);
+        }
+
+        /// <summary>
+        /// Removes all the entries from the collection.
+        /// </summary>
+        public void Clear()
+        {
+            Entries.Clear();
+        }
+
+        /// <inheritdoc />
+        public void WriteTo([NotNull] TextWriter writer, [NotNull] IHtmlEncoder encoder)
+        {
+            foreach (var entry in Entries)
+            {
+                var entryAsString = entry as string;
+                if (entryAsString != null)
+                {
+                    // encoder.HtmlEncode(entryAsString, writer);
+                    writer.Write(entryAsString);
+                }
+                else
+                {
+                    // Only string, IHtmlContent values can be added to the buffer.
+                    var entryAsIHtmlContent = (IHtmlContent)entry;
+                    if (entryAsIHtmlContent != null)
```

Comment:
We could throw here with for an entry that is a ```string``` but is null. Since we're exposing the ```List<object>``` as an internal property we don't know going forward that nobody is going to do that... perhaps the whole thing should be based on ```is string``` and ```is IHtmlContent``` rather that throwing an ```InvalidCastException```